ddtrace/pin.py

```diff
@@ -46,15 +50,23 @@ def __repr__(self):
 
     @staticmethod
     def get_from(obj):
-        """ Return the pin associated with the given object.
+        """Return the pin associated with the given object. If a pin is attached to
+        `obj` but the instance is not the owner of the pin, a new pin is cloned and
+        attached. This ensures that a pin inherited from a class is a copy for the new
+        instance, avoiding that a specific instance overrides other pins values.
 
             >>> pin = Pin.get_from(conn)
         """
         if hasattr(obj, '__getddpin__'):
             return obj.__getddpin__()
 
         pin_name = _DD_PIN_PROXY_NAME if isinstance(obj, wrapt.ObjectProxy) else _DD_PIN_NAME
-        return getattr(obj, pin_name, None)
+        pin = getattr(obj, pin_name, None)
+        # detect if the PIN has been inherited from a class
```

Comment:
This is a **caveat**. The current implementation allows to:
* set a pin in a global variable (i.e. `requests` library to set global settings)
* when an instance is created from a class that has a `Pin` attached, the pin reference is copied to the instance, meaning they both point to the same object
* because the `config` object is a mutable value, it means that changing the configuration for an "instance" pin, changes also the global configuration for all other pins
* this results in an error

All of this is done to keep the API simple, while reusing the `Pin` class. Another approach is to continue to expose and suggest the `Pin` API, that I would rather to hide / remove from the public facing API.

Reply:
Should we review our integrations to be sure this mechanism wasn't used in any of them?

Reply:
Yes it's part of the task and it seems we don't do that. Mostly because the caveat is present only for the `config` object and in the current implementation only the `tags` is affected (so it's a bugfix in that sense). Indeed, if we change a service name for an instance of the main class, we must use the `override()` method that clones the `Pin`.

This behavior protects only the `tags` and `config` object from that change, so a bugfix and a new feature. We should be good here.